libvast/builtins/operators/import.cpp

```diff
@@ -0,0 +1,154 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/concept/parseable/string/char_class.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/error.hpp>
+#include <vast/logger.hpp>
+#include <vast/pipeline.hpp>
+#include <vast/plugin.hpp>
+#include <vast/system/node_control.hpp>
+
+#include <arrow/type.h>
+#include <caf/attach_stream_source.hpp>
+#include <caf/scheduled_actor.hpp>
+#include <caf/scoped_actor.hpp>
+#include <caf/typed_event_based_actor.hpp>
+
+namespace vast::plugins::import {
+
+namespace {
+
+class import_source_driver final
+  : public caf::stream_source_driver<
+      caf::broadcast_downstream_manager<table_slice>> {
+public:
+  import_source_driver(generator<table_slice>& input, uint64_t& num_events,
+                       operator_control_plane& ctrl)
+    : input_{input}, num_events_{num_events}, ctrl_{ctrl} {
+  }
+
+  void pull(caf::downstream<table_slice>& out, size_t num) override {
+    auto it = input_.unsafe_current();
+    for (size_t i = 0; i < num; ++i) {
+      auto next = std::move(*++it);
+      if (next.rows() == 0) {
```

Comment:
What if `++it == input_.end()`?

Reply:
Good point; I think I never ran into this because num is usually 1, and then we check it after every iteration. I'll add a check to return early in that case.

Reply:
Fixed in db6a764ba3